api/src/main/java/org/openmrs/api/storage/S3/S3StorageService.java

```diff
@@ -0,0 +1,139 @@
+public S3StorageService(StreamDataService streamDataService) {
+    super(streamDataService);
+    this.s3Client = AmazonS3ClientBuilder.standard().build();
+}
+
+    @Override
+    public InputStream getData(String key) throws IOException {
+        S3Object s3object = s3Client.getObject(bucketName, key);
+        return s3object.getObjectContent();
+    }
+
+    @Override
+    public InputStream getTempData(String key) throws IOException {
+        return getData("temp/" + key);
+    }
+
+    @Override
+    public ObjectMetadata getMetadata(String key) throws IOException {
+        ObjectMetadata metadata = new ObjectMetadata();
+        com.amazonaws.services.s3.model.ObjectMetadata awsMetadata = s3Client.getObjectMetadata(bucketName, key);
+        metadata.setContentType(awsMetadata.getContentType());
+        metadata.setSize(awsMetadata.getContentLength());
+        return metadata;
+    }
+
+    @Override
+    public Stream<String> getKeys(String moduleIdOrGroup, String keyPrefix) throws IOException {
+        ListObjectsV2Request req = new ListObjectsV2Request().withBucketName(bucketName).withPrefix(moduleIdOrGroup + "/" + keyPrefix);
+        ListObjectsV2Result result = s3Client.listObjectsV2(req);
+        return result.getObjectSummaries().stream().map(S3ObjectSummary::getKey);
+    }
+
+    @Override
+    public String saveData(InputStream inputStream, ObjectMetadata metadata, String moduleIdOrGroup) throws IOException {
+        String key = moduleIdOrGroup + "/" + UUID.randomUUID().toString();
+        return saveData(inputStream, metadata, moduleIdOrGroup, key);
+    }
+
+    @Override
+    public String saveTempData(InputStream inputStream, ObjectMetadata metadata) throws IOException {
+        String key = "temp/" + UUID.randomUUID().toString();
+        return saveData(inputStream, metadata, "temp", key);
+    }
+
+    @Override
+    public String saveTempData(StreamDataWriter writer, ObjectMetadata metadata) throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+        writer.write(outputStream);
+        return saveTempData(new ByteArrayInputStream(outputStream.toByteArray()), metadata);
+    }
+
+    @Override
+    public String saveData(InputStream inputStream, ObjectMetadata metadata, String moduleIdOrGroup, String keySuffix) throws IOException {
+        com.amazonaws.services.s3.model.ObjectMetadata awsMetadata = new com.amazonaws.services.s3.model.ObjectMetadata();
+        if (metadata != null) {
+            awsMetadata.setContentType(metadata.getContentType());
+            awsMetadata.setContentLength(metadata.getSize());
+        }
+        s3Client.putObject(bucketName, keySuffix, inputStream, awsMetadata);
+        return keySuffix;
+    }
+
+    @Override
+    public String saveData(StreamDataWriter writer, ObjectMetadata metadata, String moduleIdOrGroup, String keySuffix) throws IOException {
+        ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
```

Comment:
Do not override. Base implementation uses streamService, which is more efficient in terms of memory usage.